Specifications/QIR/Callables.md

```diff
@@ -351,13 +351,14 @@ callable values:
 |---------------------------------|--------------------------------------------|-------------|
-| __quantum__rt__callable_create  | `%Callable*([4 x void (%TupleHeader*, %TupleHeader*, %TupleHeader*)*]*, %TupleHeader*)` | Initializes the callable with the provided function table and capture tuple. The capture tuple pointer should be null if there is no capture. |
-| __quantum__rt__callable_copy    | `%Callable*(%Callable*)`             | Initializes the first callable to be the same as the second callable. |
-| __quantum__rt__callable_invoke  | `void(%Callable*, %TupleHeader*, %TupleHeader*)` | Invokes the callable with the provided argument tuple and fills in the result tuple. |
+| __quantum__rt__callable_create  | `%Callable*([4 x void (%Tuple*, %Tuple*, %Tuple*)*]*, [2 x void(%Tuple*, i64)]*, %Tuple*)` | Initializes the callable with the provided function table, memory management table, and capture tuple. The memory management table pointer and the capture tuple pointer should be null if there is no capture. |
+| __quantum__rt__callable_copy    | `%Callable*(%Callable*, i1)`             | Creates a shallow copy of the callable if the alias count is larger than 0 or the second argument is `true`. Returns the given callable pointer otherwise, after increasing its reference count by 1. |
+| __quantum__rt__callable_invoke  | `void(%Callable*, %Tuple*, %Tuple*)` | Invokes the callable with the provided argument tuple and fills in the result tuple. |
 | __quantum__rt__callable_make_adjoint | `void(%Callable*)`                         | Updates the callable by applying the Adjoint functor. |
 | __quantum__rt__callable_make_controlled | `void(%Callable*)`                      | Updates the callable by applying the Controlled functor. |
-| __quantum__rt__callable_reference | `void(%Callable*)`                      | Indicates that a new reference has been added. |
-| __quantum__rt__callable_unreference | `void(%Callable*)`                      | Indicates that an existing reference has been removed and potentially releases the callable value. |
+| __quantum__rt__callable_update_reference_count | `void(%Callable*, i64)`                      | Adds the given integer value to the reference count for the callable. Deallocates the callable if the reference count becomes 0. The behavior is undefined if the reference count becomes negative. |
+| __quantum__rt__callable_update_alias_count | `void(%Callable*, i64)`                      | Adds the given integer value to the alias count for the callable. Fails if the count becomes negative. |
+| __quantum__rt__callable_memory_management | `void(i32, %Callable*, i64)`                      | Invokes the function at the given index in the memory management table of the callable with the capture tuple and the given 64-bit integer. Does nothing if if the memory management table pointer or the function pointer at that index is null.  |
```

Comment:
I'd put %Callable* argument first to be consistent with other methods in the table (and with the convention of passing 'this' pointer) but I don't feel strongly about it. Your call.

Reply:
Let's go with Alan's suggestion below and actually specify which function has to be at which index and give them distinct names.